Add circular padding for artifact-free StableDiffusionPanoramaPipeline

EvgenyKashin opened the pull request:
## Artifact-Free Panorama Generation with circular padding
This pull request addresses the issue mentioned in https://github.com/huggingface/diffusers/issues/4024. The proposed solution aims to achieve artifact-free panorama generation by seamlessly **aligning the content on the left and right sides of the generated image**, creating a proper “panorama”, viewable in 360 panoramas viewers.

In the existing MultiDiffusion approach, the image is cropped from left to right with a specified stride. To enable circular padding and support the desired 360-degree scenario, this PR introduces additional crops after the rightmost point of the image. **Circular padding allows the model to “see” the transition from the rightmost part of the image to the leftmost part, maintaining consistency in 360-sense**.

Furthermore, modifications are made to the decoding of latents in the case of StableDiffusion. Without these modifications, the decoded latents would not match in the RGB space, even if they align well in the latent space. Thus, **the decoding of latents also incorporates circular padding to ensure consistent results**.

A panorama image generated **without circular padding**: ![image](https://github.com/EvgenyKashin/diffusers/blob/images/temp_image_folder/indoor14.png?raw=true)

A panorama image generated **with circular padding**: ![image](https://github.com/EvgenyKashin/diffusers/blob/images/temp_image_folder/indoor14_circ.png?raw=true)

To maintain backward compatibility and preserve the previous behaviour, the proposed circular padding functionality is implemented as an optional argument “circular_padding”, which is set to False by default.

Usage example:
```python
import torch
from diffusers import StableDiffusionPanoramaPipeline, DDIMScheduler

model_ckpt = "stabilityai/stable-diffusion-2-base"
scheduler = DDIMScheduler.from_pretrained(model_ckpt, subfolder="scheduler")
pipe = StableDiffusionPanoramaPipeline.from_pretrained(model_ckpt, scheduler=scheduler, torch_dtype=torch.float16)
pipe = pipe.to("cuda")

prompt = "an indoor equirectangular panorama of a living room"

image = pipe(prompt, width=1024, height=512,
    circular_padding=True).images[0]
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_4025). All of your documentation changes will be reflected on that endpoint.

EvgenyKashin commented:
Visualization of a concatenation of the most right and the most left parts of a generated panorama images: No circular padding (default):
 ![image](https://github.com/EvgenyKashin/diffusers/blob/images/temp_image_folder/no_circ.png?raw=true)

Circular padding, but without decoding padding in the VAE (`decode_latents_with_padding`):
![image](https://github.com/EvgenyKashin/diffusers/blob/images/temp_image_folder/vae_no_padding.png?raw=true)

Circular padding full (`circular_padding=True`):
![image](https://github.com/EvgenyKashin/diffusers/blob/images/temp_image_folder/vae_padding.png?raw=true)

EvgenyKashin commented:
In general, it's an easy fix for the stitching artifact problem, but maybe I've missed an easier way to do it with the current pipeline. Perhaps this is too narrow case to add this to the main. 

I've also noticed [tests](https://github.com/huggingface/diffusers/blob/main/tests/pipelines/stable_diffusion/test_stable_diffusion_panorama.py) for this pipeline that I break with my changes (even with default `circular_padding=False` option, that is just the same as current main in my local tests):
`18 failed, 5 passed, 5 skipped, 2 warnings`

I've tried "cuda" and "cpu", different `batch_size` and `view_batch_size`, it works fine. I couldn't understand why it's failing tests yet.

EvgenyKashin commented:
@omerbt @sayakpaul @Isotr0py guys, what do you think?

sayakpaul commented:
I will defer to @omerbt for this one. But I like what I see :)

omerbt commented:
Looks great!

sayakpaul commented:
Thanks @omerbt!

@EvgenyKashin, let's make sure the tests on the CI pass. Then I think we can approach merging.

EvgenyKashin commented:
Thanks, guys. @omerbt I've fixed one place in the code. Basically, I decided to save one line originally, but one more check was required for small images case (like in test cases). I've also added a few simple test cases for circular padding.

EvgenyKashin commented:
Sorry, originally tagged the wrong person, @sayakpaul

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_panorama.py:
```diff
@@ -560,6 +577,8 @@ def __call__(
                 `self.processor` in
                 [diffusers.cross_attention](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/cross_attention.py).
+            circular_padding (`bool`, *optional*, defaults to `False`):
+                Whether to use circular padding for the diffusion
```
Maybe explain a bit what circular padding is and how it helps?

EvgenyKashin replied:
Done

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_panorama.py:
```diff
@@ -711,7 +760,10 @@ def __call__(
         if not output_type == "latent":
-            image = self.vae.decode(latents / self.vae.config.scaling_factor, return_dict=False)[0]
+            if circular_padding:
+                image = self.decode_latents_with_padding(latents)
```
Works for me for now. Ideally, I would have loved to move out the VAE decode part from `decode_latents_with_padding()` to unify this part of the workflow a bit. But okay for now.

sayakpaul commented on tests/pipelines/stable_diffusion/test_stable_diffusion_panorama.py:
```diff
@@ -260,6 +294,36 @@ def test_stable_diffusion_panorama_default(self):
 
         assert np.abs(expected_slice - image_slice).max() < 1e-2
 
+    def test_stable_diffusion_panorama_circular_padding(self):
```
I think it's okay to remove this slow test.

EvgenyKashin replied:
Removed

sayakpaul reviewed (approved):
Looks great to me! Thank you. 

Left some minor nits. One thing we could do is add a note about `circular_padding` to the panorama docs: https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/pipelines/panorama.mdx.

EvgenyKashin commented:
Could we merge it?

sayakpaul commented on docs/source/en/api/pipelines/panorama.mdx:
```diff
@@ -60,6 +60,19 @@ and increase the VRAM usage.
 
+<Tip>
+
+Circular padding is applied to ensure there are no stitching artifacts when working with 
+panoramas that needs to seamlessly transition from the rightmost part to the leftmost part. 
+By enabling circular padding (set `circular_padding=True`), the operation applies additional 
+crops after the rightmost point of the image, allowing the model to "see” the transition 
+from the rightmost part to the leftmost part. This helps maintain visual consistency in 
+a 360-degree sense and creates a proper “panorama” that can be viewed using 360-degree 
+panorama viewers. When decoding latents in StableDiffusion, circular padding is applied 
+to ensure that the decoded latents match in the RGB space.
```
If possible would be also nice to link the comparison with and without circular padding. This way, the readers can spot-check the effectivity.

EvgenyKashin replied:
Do you mean adding a text description of “normal” multi-diffusion, or adding a link to an image example of the results?

sayakpaul replied:
> adding a link to an image example of the results?

Yeah just noting the difference with a comparative example like the one you showed in the PR description.

EvgenyKashin replied:
That was a good idea indeed

sayakpaul commented:
Will merge after the CI passes.

sayakpaul commented:
Thanks very much for your valuable contributions!
